### ISSUE TITLE
Clang does not recognize portable add-with-carry patterns

### ISSUE THREAD

**Issue by davidben** [llvm/llvm-project#73847]
When writing code for cryptographic primitives, or big integers in general, one often needs to use the ISA's add-with-carry instructions and chain carry flags up a bignum.

Although Clang does provide Clang-specific intrinsics like `__builtin_addc`, they're not portable across compilers. I made several attempts to write a portable add-with-carry here, and Clang seems unable to recognize any of them. Here's a godbolt link with a bunch of them:
https://godbolt.org/z/WTns6M8E6

(CC @andres-erbsen, do you remember if there were other patterns we'd tried?)

**Comment by davidben** [comment]
@EugeneZelenko are you sure `clang:headers` is the right label? This isn't an issue with the intrinsics, but that very small snippets of portable code are optimized in the same way as the intrinsics.

**Comment by shafik** [comment]
@AaronBallman @erichkeane any advice here?

**Comment by erichkeane** [comment]
This is likely something that the opt folks have to take a look at.  @topperc was particularly good at bit pattern recognition at one point, so he might be able to help out.

**Comment by topperc** [comment]
FYI gcc does support __builtin_addc so it isn't clang specific anymore.

Do other compilers optimize any of your sequences?

**Comment by davidben** [comment]
> FYI gcc does support __builtin_addc so it isn't clang specific anymore.

Yeah, I realized that after I filed this. We're using that when available, annoying as the ifdef soup is. :-) But it'd be nice if more portable sequences were recognized so not every project needs to discover this on their own.

> Do other compilers optimize any of your sequences?

Not that I'm aware of. The compilers we've used seem to be uniformly pretty bad at handling carry flags, alas. This was filed less as a feature parity thing and more as a missed optimization. (This sort of code, when really perf-sensitive, often needs to dip all the way into assembly to avoid compiler mishaps, so there's a long way to code.)
